lib/gc/Dialect/Linalgx/Utils.cpp

```diff
@@ -0,0 +1,529 @@
+                            SmallVector<SmallVector<AffineExpr>> &exprsArr,
+                            ShapedType shapeA, ShapedType shapeB,
+                            ShapedType shapeC, const PackingAttr &attr) {
+  SmallVector<AffineExpr> exprsA(shapeA.getRank());
+  SmallVector<AffineExpr> exprsB(shapeB.getRank());
+  SmallVector<AffineExpr> exprsC(shapeC.getRank());
+  unsigned dims = 0;
+  // dims count from 0
+  auto getBatchExprs = [&](const BatchDimMap &batchDimMap) {
+    for (; (int64_t)dims < batchDimMap.getBatchNum(); dims++) {
+      auto curr = getAffineDimExpr(dims, context);
+      exprsA[batchDimMap.getBatchA()[dims]] = curr;
+      exprsB[batchDimMap.getBatchB()[dims]] = curr;
+      if (batchDimMap.isBatchMatmul())
+        exprsC[batchDimMap.getBatchC()[dims]] = curr;
+    }
+  };
+  auto getPackingExprs = [&](ArrayRef<PackingMap> mapArray,
+                             ArrayRef<ShapedType> types,
+                             ArrayRef<SmallVector<AffineExpr> *> exprs) {
+    for (auto &packingMap : mapArray) {
+      unsigned srcIndex = packingMap.getPackingSrcIndex();
+      unsigned dstIndex = packingMap.getPackingDstIndex();
+      ArrayRef<int64_t> srcDims = packingMap.getPackingSrcDims();
+      ArrayRef<int64_t> dstDims = packingMap.getPackingDstDims();
+      SmallVector<AffineExpr> &dstExprs = *exprs[dstIndex];
+      SmallVector<AffineExpr> &srcExprs = *exprs[srcIndex];
+      AffineExpr compound = getAffineConstantExpr(0, context);
+      for (auto dim : dstDims) {
+        AffineExpr curr = getAffineDimExpr(dims++, context);
+        AffineExpr constant =
+            getAffineConstantExpr(types[dstIndex].getDimSize(dim), context);
+        compound = compound * constant + curr;
+        dstExprs[dim] = curr;
+      }
+      srcExprs[srcDims.front()] = compound;
+    }
+  };
+  // Process order: b, m, n, k packing, kept same as packing iterator types
+  getBatchExprs(attr.batchDimMap);
+  getPackingExprs(attr.mPacking, {shapeA, shapeC}, {&exprsA, &exprsC});
+  getPackingExprs(attr.nPacking, {shapeB, shapeC}, {&exprsB, &exprsC});
+  getPackingExprs(attr.kPacking, {shapeA, shapeB}, {&exprsA, &exprsB});
+  exprsArr.emplace_back(exprsA);
+  exprsArr.emplace_back(exprsB);
+  exprsArr.emplace_back(exprsC);
+  return dims;
+}
+
+SmallVector<AffineMap> getIndexingMaps(MLIRContext *context, ShapedType shapeA,
+                                       ShapedType shapeB, ShapedType shapeC,
+                                       const PackingAttr &attr) {
+  SmallVector<SmallVector<AffineExpr>> exprsArr;
+  auto dims = getPackingDimsExpr(context, exprsArr, //
+                                 shapeA, shapeB, shapeC, attr);
+  auto mapA = simplifyAffineMap(AffineMap::get(dims, 0, exprsArr[0], context));
+  auto mapB = simplifyAffineMap(AffineMap::get(dims, 0, exprsArr[1], context));
+  auto mapC = simplifyAffineMap(AffineMap::get(dims, 0, exprsArr[2], context));
+  return {mapA, mapB, mapC};
```

Comment:
There are too many `auto` here. Please follow this [COC](https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable) of LLVM.